File: src/core/logger.cpp
#include <unordered_map>
#include <spdlog/sinks/stdout_color_sinks.h>
#include "magnetar/core/logger.h"

magnetar::LogLevel magnetar::Logger::s_level = magnetar::LogLevel::debug;

namespace magnetar
{
    static std::vector<const char *> s_tag_list = {
        logger::tags::application,
        logger::tags::events,
        logger::tags::input,
        logger::tags::renderer,
        logger::tags::assets,
    };
}
void magnetar::Logger::initialize(LogLevel level)
{
    s_level = level;
    spdlog::default_logger()->set_level(level);

    for (auto &tag : s_tag_list)
    {
        auto logger = spdlog::stdout_color_mt(tag);
        logger->set_level(level);
    }
}

void magnetar::Logger::set_level(LogLevel level)
{
    spdlog::default_logger()->set_level(level);
    for (auto &tag : s_tag_list)
        spdlog::get(tag)->set_level(level);
}
void magnetar::Logger::add_tag(const std::string &name)
{
    auto logger = spdlog::stdout_color_mt(name);
    logger->set_level(s_level);
}

std::shared_ptr<spdlog::logger> magnetar::Logger::get(const char *name)
{
    if (!name)
        return spdlog::default_logger();
    return spdlog::get(name);
}

File: Magnetar/Core/Logger.cs

namespace Magnetar.Core
{
    public static class Logger
    {
        public static void Fatal(string message)
        {
            InternalCalls.Logger_Log(LogLevel.Fatal, message);
        }

        public static void Error(string message)
        {
            InternalCalls.Logger_Log(LogLevel.Error, message);
        }

        public static void Warn(string message)
        {
            InternalCalls.Logger_Log(LogLevel.Warn, message);
        }

        public static void Info(string message)
        {
            InternalCalls.Logger_Log(LogLevel.Info, message);
        }

        public static void Debug(string message)
        {
            InternalCalls.Logger_Log(LogLevel.Debug, message);
        }

        public static void Trace(string message)
        {
            InternalCalls.Logger_Log(LogLevel.Trace, message);
        }

        public static void Log(LogLevel level, string message)
        {
            InternalCalls.Logger_Log(level, message);
        }
    }
}